test/KQStream.ts

```diff
@@ -42,142 +39,134 @@ function removeAllListeners(stream: KQStream) {
-            playerNamesPromise = createListenerPromise(stream, 'playernames', onPlayerNames, 1);
-            playerKillPromise = createListenerPromise(stream, 'playerKill', onPlayerKill, 4);
-
-            // Act
-            for (let event of events) {
-                cab.send(event.message);
+            for (let message of messages) {
+                await new Promise<void>((resolve) => {
+                    stream.once(message.type, (arg) => {
+                        (receivedEvents[message.type] as any).push(arg);
+                        resolve();
+                    });
+                    cab.send(message.data);
+                });
             }
         });
 
         it('should process playernames events', async () => {
-            await playerNamesPromise;
-            expect(playerNamesEvents.length).to.equal(1);
-            for (let expectedEvent of expectedPlayerNamesEvents) {
-                expect(playerNamesEvents).to.deep.include(expectedEvent);
+            expect(receivedEvents.playernames.length).to.equal(1);
+            for (let expectedEvent of expectedEvents.playernames) {
+                expect(receivedEvents.playernames).to.deep.include(expectedEvent);
             }
         });
 
         it('should process playerKill events', async () => {
-            await playerKillPromise;
-            expect(playerKillEvents.length).to.equal(4);
-            for (let expectedEvent of expectedPlayerKillEvents) {
-                expect(playerKillEvents).to.deep.include(expectedEvent);
+            expect(receivedEvents.playerKill.length).to.equal(4);
+            for (let expectedEvent of expectedEvents.playerKill) {
+                expect(receivedEvents.playerKill).to.deep.include(expectedEvent);
             }
         });
-
-        after(() => {
-            removeAllListeners(stream);
-        });
     });
 
     describe('#read', () => {
-        before(() => {
-            // Arrange
-            playerNamesEvents = [];
-            playerKillEvents = [];
-            playerNamesPromise = createListenerPromise(stream, 'playernames', onPlayerNames, 1);
-            playerKillPromise = createListenerPromise(stream, 'playerKill', onPlayerKill, 4);
+        before(async () => {
+            receivedEvents = {
+                playernames: [],
+                playerKill: []
+            };
+
+            stream.on('playernames', (arg) => {
+                receivedEvents.playernames.push(arg);
+            });
+            stream.on('playerKill', (arg) => {
+                receivedEvents.playerKill.push(arg);
+            });
+
             let data: string = '';
-            for (let event of events) {
-                data += `${event.timestamp},${event.message}\n`;
+            for (let message of messages) {
+                data += `${message.timestamp},${message.data}\n`;
             }
-
-            // Act
             stream.read(data);
+
+            await sleep(100);
```

Comment:
This is code smell. Find a better way to know that messages have arrived.

Reply:
Also, consider deprecating `KQStream#read`. One can make the argument that KQStream is designed to only receive messages from a Killer Queen cab via websocket connection. If this is the case, why are we letting people read in some alternative format?

Using `KQCab` is the alternative, but it does add an additional layer of complexity by creating a websocket server. However, this is how a connection to a real Killer Queen cabinet would be, and isn't that what we want our code to work with? Isn't using `#read` misleading as it may cause us to overlook issues with `#connect`?

But then again, that's why we unit test both methods...

Give it some consideration.